src/proc.ts

```diff
@@ -71,10 +72,70 @@ export interface Subprocess {
 export interface BuildCommand {
   command: string;
   args?: string[];
-  build_env?: {[key: string]: string};
+  build_env?: EnvironmentVariables;
 }
 
-export interface EnvironmentVariables { [key: string]: string }
+export class EnvironmentVariables {
+  private env: {[key: string]: string } = {};
+
+  static create(env?: {[key: string]: string|null}): EnvironmentVariables;
+  static create(env?: {[key: string]: string|undefined}): EnvironmentVariables;
+  static create(env?: {[key: string]: string|null|undefined}): EnvironmentVariables {
+    const result = new EnvironmentVariables();
+    result.merge(env);
+    return result;
+  }
+
+  public get fullEnvironment() {
+    return this.env;
+  }
+
+  public get debugConfigEnvironment() {
+    return Object.entries(this.env).map(
+      ([key, value]) => ({
+        name: key,
+        value
+      })
+    );
+  }
+
+  public get(key: string): string | undefined {
+    return this.env[util.normalizeEnvironmentVarname(key)];
+  }
+
+  public set(key: string, value?: string|null): void {
+    if (value !== undefined && value !== null) {
+      this.env[util.normalizeEnvironmentVarname(key)] = value;
```

Comment:
As on Windows, we need preserve the original case when genearting 
the final environmentMap to use

Reply:
On Windows it shouldn't matter what the casing is, should it?

But if the goal is to preserve the casing, an additional object should be added that maps the normalized keys to the actual keys in the `this.env` object.  This would avoid us having to loop through all keys every time you look up a value.  That was my biggest concern with the original PR.

Reply:
e.g. add a `this.keys`

Reply:
I pushed a change that should handle the casing (though none of this code is tested yet as it is incomplete)